src/cpr_data_access/models.py

```diff
@@ -1097,3 +1105,105 @@ def to_huggingface(
         )
 
         return huggingface_dataset
+
+    def from_huggingface(
+        self,
+        huggingface_dataset: HFDataset,
+        limit: Optional[int] = None,
+    ) -> "Dataset":
+        """
+        Create a dataset from a huggingface dataset.
+
+        :param huggingface_dataset: created using `Dataset.to_huggingface()`
+        :param limit: optionally limit the number of documents to load
+        :return self: with documents loaded from huggingface dataset
+        """
+
+        document_metadata_model = self.document_model.__fields__[
+            "document_metadata"
+        ].outer_type_
+        hf_dataframe: pd.DataFrame = huggingface_dataset.to_pandas()
+
+        if limit is not None:
+            doc_ids = hf_dataframe["document_id"].unique()[:limit]
+            hf_dataframe = hf_dataframe[hf_dataframe["document_id"].isin(doc_ids)]
+
+        documents = []
+
+        for _, doc_df in tqdm(
+            hf_dataframe.groupby("document_id"),
+            total=hf_dataframe["document_id"].nunique(),
+            unit="docs",
+        ):
+            document_text_blocks = [
+                TextBlock(
+                    # TODO: we aren't able to access the original text split over lines as it's not stored in the huggingface dataset
+                    text=[row["text"]],
+                    text_block_id=row["text_block_id"],
+                    language=row["language"],
+                    type=row["type"],
+                    type_confidence=row["type_confidence"],
+                    page_number=row["page_number"],
+                    coords=[tuple(c) for c in row["coords"].tolist()] if row["coords"] is not None else None,  # type: ignore
+                )
+                for _, row in doc_df.iterrows()
+            ]
+
+            # pandas to_dict() stores sequences as numpy arrays, so we need to convert them back to lists
+            doc_fields = {
+                k: list(v) if isinstance(v, np.ndarray) else v
+                for k, v in doc_df.iloc[0].to_dict().items()
+            }
+
+            doc_metadata = document_metadata_model.parse_obj(
+                doc_fields
+                | {"source": "GST" if self.document_model == GSTDocument else "CPR"}
+            )
+
+            doc = self.document_model.parse_obj(
+                doc_fields
+                | {
+                    "document_metadata": doc_metadata,
+                    "text_blocks": document_text_blocks,
+                }
+            )
```

Comment:
Related to my question above the `__fields__`: it feels to me that the metadata parsing should be part of document parsing, as the metadata is an attribute of the doc -- this uncoupling might become hard to maintain. Is there any reason against moving the metadata parsing logic into the `parse_obj()` of the document class?

Reply:
Also, I've just realised that we're storing `doc_fields` both on the metadata and the data model itself -- is there any reason for this redundancy?

Reply:
I've refreshed but haven't seen the changes -- is that pushed already?

Reply:
yep, see [this commit](https://github.com/climatepolicyradar/data-access/pull/45/commits/ae1a689d241d181a314d41d59c9a73c03a416e6d)